File: src/main/java/destiny/core/chinese/ziwei/StarYearFront.java
/**
 * Created by smallufo on 2017-06-17.
 */
package destiny.core.chinese.ziwei;

import destiny.core.chinese.Branch;
import org.jooq.lambda.function.Function2;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * 歲前 12 星
 */
public class StarYearFront extends ZStar {

  public final static StarYearFront 歲建 = new StarYearFront("歲建");
  public final static StarYearFront 晦氣 = new StarYearFront("晦氣");
  public final static StarYearFront 喪門 = new StarYearFront("喪門");
  public final static StarYearFront 貫索 = new StarYearFront("貫索");
  public final static StarYearFront 官符 = new StarYearFront("官符");
  public final static StarYearFront 小耗 = new StarYearFront("小耗");
  public final static StarYearFront 歲破 = new StarYearFront("歲破");
  public final static StarYearFront 龍德 = new StarYearFront("龍德");
  public final static StarYearFront 白虎 = new StarYearFront("白虎");
  public final static StarYearFront 天德 = new StarYearFront("天德");
  public final static StarYearFront 吊客 = new StarYearFront("吊客");
  public final static StarYearFront 病符 = new StarYearFront("病符");

  public final static StarYearFront[] values = {歲建, 晦氣, 喪門, 貫索, 官符, 小耗, 歲破, 龍德, 白虎, 天德, 吊客, 病符};

  private final static List<StarYearFront> list = Arrays.asList(values);

  public StarYearFront(String nameKey) {
    super(nameKey, StarYearFront.class.getName(), Type.歲前);
  }

  private final static Function2<Branch , StarYearFront , Branch> fun = (年支 , 星) -> {
    int steps = list.indexOf(星);
    return 年支.next(steps);
  };

  public final static Function<Branch , Branch> fun歲建 = (年支) -> fun.apply(年支 , 歲建);
  public final static Function<Branch , Branch> fun晦氣 = (年支) -> fun.apply(年支 , 晦氣);
  public final static Function<Branch , Branch> fun喪門 = (年支) -> fun.apply(年支 , 喪門);
  public final static Function<Branch , Branch> fun貫索 = (年支) -> fun.apply(年支 , 貫索);
  public final static Function<Branch , Branch> fun官符 = (年支) -> fun.apply(年支 , 官符);
  public final static Function<Branch , Branch> fun小耗 = (年支) -> fun.apply(年支 , 小耗);
  public final static Function<Branch , Branch> fun歲破 = (年支) -> fun.apply(年支 , 歲破);
  public final static Function<Branch , Branch> fun龍德 = (年支) -> fun.apply(年支 , 龍德);
  public final static Function<Branch , Branch> fun白虎 = (年支) -> fun.apply(年支 , 白虎);
  public final static Function<Branch , Branch> fun天德 = (年支) -> fun.apply(年支 , 天德);
  public final static Function<Branch , Branch> fun吊客 = (年支) -> fun.apply(年支 , 吊客);
  public final static Function<Branch , Branch> fun病符 = (年支) -> fun.apply(年支 , 病符);

  public final static Map<StarYearFront , Function<Branch , Branch>> funMap = new HashMap<StarYearFront , Function<Branch, Branch>>() {{
    put(歲建 , fun歲建);
    put(晦氣 , fun晦氣);
    put(喪門 , fun喪門);
    put(貫索 , fun貫索);
    put(官符 , fun官符);
    put(小耗 , fun小耗);
    put(歲破 , fun歲破);
    put(龍德 , fun龍德);
    put(白虎 , fun白虎);
    put(天德 , fun天德);
    put(吊客 , fun吊客);
    put(病符 , fun病符);
  }};
}

File: src/main/java/destiny/core/chinese/ziwei/StarYearFront.kt
/**
 * Created by smallufo on 2017-06-17.
 */
package destiny.core.chinese.ziwei

import destiny.core.IPoints
import destiny.core.chinese.Branch
import java.util.*
import kotlin.reflect.KClass

/**
 * 歲前 12 星
 */
sealed class StarYearFront(nameKey: String) : ZStar(nameKey, StarYearFront::class.java.name, Type.歲前) {

  object 歲建 : StarYearFront("歲建")
  object 晦氣 : StarYearFront("晦氣")
  object 喪門 : StarYearFront("喪門")
  object 貫索 : StarYearFront("貫索")
  object 官符 : StarYearFront("官符")
  object 小耗 : StarYearFront("小耗")
  object 歲破 : StarYearFront("歲破")
  object 龍德 : StarYearFront("龍德")
  object 白虎 : StarYearFront("白虎")
  object 天德 : StarYearFront("天德")
  object 吊客 : StarYearFront("吊客")
  object 病符 : StarYearFront("病符")

  override fun compareTo(other: ZStar): Int {
    return if (other is StarYearFront) {
      values.indexOf(this) - values.indexOf(other)
    } else {
      super.compareTo(other)
    }
  }

  companion object : IPoints<StarYearFront> {

    override val type: KClass<out ZStar> = StarYearFront::class

    override val values by lazy { arrayOf(歲建, 晦氣, 喪門, 貫索, 官符, 小耗, 歲破, 龍德, 白虎, 天德, 吊客, 病符) }

    override fun fromString(value: String, locale: Locale): StarYearFront? {
      return values.firstOrNull {
        it.nameKey == value
      }
    }

    private val list by lazy { listOf(*values) }

    private val func = { yearBranch: Branch, star: ZStar ->
      val steps = list.indexOf(star)
      yearBranch.next(steps)
    }

    val fun歲建 = { yearBranch: Branch -> func.invoke(yearBranch, 歲建) }
    val fun晦氣 = { yearBranch: Branch -> func.invoke(yearBranch, 晦氣) }
    val fun喪門 = { yearBranch: Branch -> func.invoke(yearBranch, 喪門) }
    val fun貫索 = { yearBranch: Branch -> func.invoke(yearBranch, 貫索) }
    val fun官符 = { yearBranch: Branch -> func.invoke(yearBranch, 官符) }
    val fun小耗 = { yearBranch: Branch -> func.invoke(yearBranch, 小耗) }
    val fun歲破 = { yearBranch: Branch -> func.invoke(yearBranch, 歲破) }
    val fun龍德 = { yearBranch: Branch -> func.invoke(yearBranch, 龍德) }
    val fun白虎 = { yearBranch: Branch -> func.invoke(yearBranch, 白虎) }
    val fun天德 = { yearBranch: Branch -> func.invoke(yearBranch, 天德) }
    val fun吊客 = { yearBranch: Branch -> func.invoke(yearBranch, 吊客) }
    val fun病符 = { yearBranch: Branch -> func.invoke(yearBranch, 病符) }

    val starFuncMap: Map<StarYearFront, Function1<Branch, Branch>> by lazy {
      mapOf(
        歲建 to fun歲建,
        晦氣 to fun晦氣,
        喪門 to fun喪門,
        貫索 to fun貫索,
        官符 to fun官符,
        小耗 to fun小耗,
        歲破 to fun歲破,
        龍德 to fun龍德,
        白虎 to fun白虎,
        天德 to fun天德,
        吊客 to fun吊客,
        病符 to fun病符
           )
    }
  }
}
